Document roles and permissions (#268)
owncloud/docs-ocis: modules/ROOT/pages/deployment/general/roles-permissions.adoc

Examples where the frontent currently provides limited capabilities to users though the backend has more capabilities:

* Admins, Space Admins and Users do not have full configuration and managability via the web interface.
* Admins and Space Admins can currently only view Spaces that have been created by them selves.
* Users cant create Spaces on their own.

Over time, methods will be implemented in the frontend to provide capabilities the backend provides in the best and most user friendly way.

== Backend Roles and Permissions

Currently there are four predefined backend roles which have fixed assigned permissions where each permission has grants specific to a role. To be noted, that some grants are given only for the account itself and not in general.

Consider that these roles and permissions are provided by the backend via API's which means you can always call them by own code or via `curl`. 

Read the image below the following way::

> * A Role gets one or more permission types assigned

Reviewer: ```suggestion
* A Role gets one or more permission types assigned.
```

> * Each permission type is granted individual rights relevant for that role

Reviewer: ```suggestion
* Each permission type is granted individual rights relevant for that role.
```

> * Roles are assigned to accounts

Reviewer: ```suggestion
* Roles are assigned to accounts.
```

// the empty macro is necessary to have more distance between text and image
 
{empty} +

image::deployment/general/roles-and-permissions.svg[width=700]

> === Backend Role Details

Reviewer: ```suggestion
=== Back-end Role Details
```

> In general with regards to Spaces, these can always be deleted by the creator and do not need a special permission assignment with the exeption of the `Delete Home Space` permission, which is only assigned to the Admin.

Reviewer: ```suggestion
In general with regards to spaces, these can always be deleted by the creator and do not need a special permission assignment with the exception of the `Delete Home Space` permission, which is only assigned to the Admin.
```

Admin::

> The `Admin` role is in full control of all permission types with one exeption which is `Self Management`. This permission type is excluded to avoid that the last admin accidentially reduces his own role to a lower one making the system unmanagable. As long there is more than one admin, roles can be set by the `Role Management` permission.

Reviewer: ```suggestion
The `Admin` role is in full control of all permission types with one exception which is `Self Management`. This permission type is excluded to avoid that the last admin accidentally reduces their own role to a lower one making the system unmanageable. As long as there is more than one admin, roles can be set by the `Role Management` permission.
```

Space Admin::

> The `Space Admin` role is capable to manage spaces and set quotas.

Reviewer: ```suggestion
The `Space Admin` role allows you to manage spaces and set quotas.
```

User::

> The `User` role can manage his own account, his language and the spaces he created but is not capable to set quotas for his spaces.

Reviewer: ```suggestion
With the `User` role you can manage your own account, change the language and edit spaces you created but you are not able to set quotas for your spaces.
```

Guest::

> The `Guest` role can only manage his language but no other settings.

Reviewer: ```suggestion
Users with the `Guest` role assigned can only change the language but no other settings.
```

> == Frontend Roles and Permissions

Reviewer: ```suggestion
== Front-end Roles and Permissions
```

////
=== Viewer

> Users with the role _viewer_ for a certain space can only

Reviewer: ```suggestion
Users with the role _Viewer_ for a certain space can only
```

* read documents,
* view images,
* play videos, etc.,

but not change anything.

=== Editor

_Editors_ of a space have additional permissions. They can:

* modify documents and even delete them,
* upload content to a space,
* edit the space name,
* edit the space description,
* add or change the image for the space,
* empty the trash bin of the space.

=== Manager

> In addition to the viewer and editor permissions, the _manager_ can:

Reviewer: ```suggestion
In addition to the viewer and editor permissions, the _Manager_ can:
```

* invite additional members,
* edit the quota of a space,
* change the roles of members, including those of other managers. A downgrade to viewer or editor is possible. 

The creator of a space becomes manager and can delegate this role to another member.

> // Unclear if only Admin can create spaces or everyone.

Reviewer: ```suggestion
```

Space Admin::
The `Space Admin` role allows you to manage spaces and set quotas.

User::
With the `User` role you can manage your own account, change the language and edit spaces you created but you are not able to set quotas for your spaces.

Guest::
Users with the `Guest` role assigned can only change the language but no other settings.

== Front-end Roles and Permissions
////
=== Viewer

Users with the role _Viewer_ for a certain space can only

* read documents,
* view images,
* play videos, etc.,

but not change anything.

=== Editor

_Editors_ of a space have additional permissions. They can:

* modify documents and even delete them,
* upload content to a space,
* edit the space name,
* edit the space description,
* add or change the image for the space,
* empty the trash bin of the space.

=== Manager

In addition to the viewer and editor permissions, the _Manager_ can:

* invite additional members,
* edit the quota of a space,
* change the roles of members, including those of other managers. A downgrade to viewer or editor is possible. 

The creator of a space becomes manager and can delegate this role to another member.

=== Admin

> The _admin_ user is created during the setup of on Infinite Scale instance. Check out the ownCloud Web section of the documentation for details on all configuration options for admins.

Reviewer 2: ```suggestion
The _admin_ user is created during the setup of an Infinite Scale instance. Check out the ownCloud Web section of the documentation for details on all configuration options for admins.
```

> With regard to spaces, roles and permissions, the admin user can create a new space and is then automatically assigned the manager role. After inviting more members to a space, the manager role can be delegated to another person. If an admin then downgrades his or her role to editor or viewer, this role can only be changes by the manager. The admin has no special privileges with regard to spaces.

Reviewer 2: ```suggestion
With regard to spaces, roles and permissions, the admin user can create a new space and is then automatically assigned the manager role. After inviting more members to a space, the manager role can be delegated to another person. If an admin then downgrades his or her role to editor or viewer, this role can only be changed by the manager. The admin has no special privileges with regard to spaces.
```

= Roles and Permissions
:toc: right

> :description: With Infinite Scale, roles and permissions are separated from a system, space and sharing point of view. The way this has been implemented makes that concept not only secure but open for upcoming and extended features.

Reviewer: ```suggestion
:description: With Infinite Scale, roles and permissions are different when it comes to the system, spaces and sharing. The implementation of this concept provides security and openness for upcoming and extended features.
```

== Introduction

{description}  

== General Info

[NOTE]
====
* To avoid confusing the role `User` with the account of a user who gets a role assigned, capitalization is used. When using `User`, the system role is meant, while `user` refers to the account.

> Roles and permissions in this document refer to users but *not* in the context of sharing.

Reviewer: ```suggestion
* Roles and permissions in this document refer to users but *not* in the context of sharing.
```

> * Roles and permissions in this document are in context of users but *not* in context of sharing.

Author: > ```suggestion
> 
> Roles and permissions in this document refer to users but *not* in the context of sharing.
> 
> ```

You remove the star making the line a list item, but the line above is still a list item? What is the intention of this?

====

=== Role Models

> Infinite Scale separates between global system roles and access roles for spaces. With global system roles, Infinite Scale grants permissions for the handling and lifecycle of tasks and spaces, where these permissions are *not* connected to access space content or further managing the space. The only exeption is, when a space gets created, the creator is also granted the Manager role of the space - which can, and usually is delegated.

Reviewer: ```suggestion
Infinite Scale distinguishes between global system roles and access roles for spaces. With global system roles, Infinite Scale grants permissions for the handling and lifecycle of tasks and spaces, where these permissions are *not* connected to access space content or further managing the space. The only exception occurs, when a space is created. Then the creator is also granted the Manager role for the space, which can be and usually is delegated.
```

> Infinite Scale distinguishes between global system roles and access roles for spaces. With global system roles, Infinite Scale grants permissions for the handling and lifecycle of tasks and spaces, where these permissions are *not* connected to access space content or further managing the space. The only exception occurs, when a space is created. Then the creator is also granted the Manager role for the space, which can be and usually is delegated.

Reviewer 2: ```suggestion
Infinite Scale distinguishes between global system roles and access roles for spaces. With global system roles, Infinite Scale grants permissions for the handling and lifecycle of tasks and spaces, where these permissions are *not* connected to accessing space content or further managing the space. The only exception occurs, when a space is created. Then the creator is also granted the Manager role for the space, which can be and usually is delegated.
```

While system roles are stored with the user as meta data, space roles are stored as extended attributes in the space.

> Currently, ownCloud web does not provide all capabilities from the backend to users as not everything has been developed so far. Therefore you will see some restrictions to protect the system respectively to avoid confusing users.

Reviewer: ```suggestion
Currently, ownCloud web does not provide all capabilities available in the backend as not everything has been fully developed so far. Therefore you will see some restrictions to protect the system and to avoid confusing users.
```

> Examples where the web front-end currently provides limited capabilities to users though the system has more capabilities:

Reviewer: ```suggestion
Examples where the front-end web interface currently provides limited features to users although the system has more capabilities:
```

* Admins, Space Admins and Users do not have full configuration and management options via the web interface.

> * Admins and Space Admins can currently only view and manage spaces they are member of.

Reviewer: ```suggestion
* Admins and Space Admins can currently only view and manage spaces they are members of.
```

* Users can't create spaces or set quotas for spaces they manage.

> Over time, the web front-end will be updated to provide capabilities in the best and most user friendly way.

Reviewer: ```suggestion
Over time, the web interface will be updated to provide capabilities in the best and most user-friendly way.
```

=== Permission Granularity

> Compared to ownCloud Server where e.g. the administrator has more or less full access to everything, Infinite Scale refines roles with assigned permissions types and particular permissions for that type on a much more detailed level. This is not only true for system roles abut also for spaces roles.

Reviewer: ```suggestion
Compared to ownCloud Server where e.g. the administrator has more or less full access to everything, Infinite Scale refines roles with assigned permission types and particular permissions for each type on a much more detailed level. This is not only true for system roles but also for space roles.
```

=== Security

> With role and permission separation (system/spaces/sharing), the overall security is greatly increased and security standards like *separating tasks and data access by design* is addressed.

Reviewer: ```suggestion
With role and permission separation (system/spaces/sharing), the overall security is greatly increased and security standards like *separating tasks and data access by design* are addressed.
```

== System Roles and Permissions

> Currently there are four predefined system roles with fixed permissions assigned to specific roles. Note that some permissions are only granted via the role for the particular user but not covering all users. A defined role can be assigned to a user, concatenating roles is currently not supported.

Reviewer: ```suggestion
Currently there are four predefined system roles with fixed permissions. Via these roles, permissions are only granted to a particular user and not all users. Concatenating roles is currently not supported.
```

Consider that these roles and permissions are provided by the system via APIs, which means you can call them via own code or `curl`. 

Read the image below the following way::
* A Role gets one or more permission types assigned.

> * Each permission type is granted individual rights relevant for that role.

Reviewer: ```suggestion
* Each permission type grants individual rights relevant for that role.
```

* Roles are assigned to users.

// the empty macro is necessary to have more distance between text and image
 
{empty} +

> // the table references following ocis code where roles and permissions a re defined: https://github.com/owncloud/ocis/blob/master/services/settings/pkg/store/defaults/defaults.go

Reviewer 2: ```suggestion
// the table references following ocis code where roles and permissions are defined: https://github.com/owncloud/ocis/blob/master/services/settings/pkg/store/defaults/defaults.go
```

image::deployment/general/roles-and-permissions-matrix.svg[width=550]

=== System Role Details

> In general with regards to spaces:

Reviewer: ```suggestion
In general with regards to spaces, the following applies:
```

> * Spaces can always be deleted by the those who are granted managing the space with the exception of the `Delete Home Space` permission, which is only assigned to the Admin for the reason when a user account gets deleted.

Reviewer: ```suggestion
* Spaces can always be deleted by space managers - with the exception of the `Delete Home Space` permission, which is only assigned to the Admin since this should only happen when a user account gets deleted.
```

* Managing a space means creating or deleting a space, setting quotas, delegating the right to manage space members and space member roles, but not having access to the content of the space except when being a member of the space.
* When creating a space, the creator is automatically in the delegatable Manager role for this space. For details see xref:space-roles-and-permissions[Space Roles and Permissions].

Admin::

> The `Admin` role is in full control of all permission types with one exception which is `Self Management`. This permission type is excluded to avoid that the last admin accidentally reduces their own role to a lower one making the system unmanageable. As long as there is more than one admin, roles can be set by the `Role Management` permission.

Reviewer: ```suggestion
The `Admin` role provides full control of all permission types with one exception, which is `Self Management`. This permission type is excluded to avoid that the last remaining admin accidentally reduces their own role to a lower one making the system unmanageable. As long as there is more than one admin, roles can be set via the `Role Management` permission.
```

Space Admin::
The `Space Admin` role allows you to create spaces, set quotas and delegating the right to manage space members and space member roles.

User::

> If you have the `User` role assigned, you can manage your own account and change the language. To manage space group members and space member roles, an Admin or Space Admin must create one and grant the Manager permission to the user.

Reviewer: Do space groups exist? What does the (Space) Admin need to create?

Author: good point, need to check this

> If you have the `User` role assigned, you can manage your own account and change the language. To manage space group members and space member roles, an Admin or Space Admin must create one and grant the Manager permission to the user.

Reviewer: ```suggestion
If you have the `User` role assigned, you can manage your own account and change the language.
```

> The `User` role can manage his own account and change his language. To manage space group members and space member roles, an Admin or Space Admin must create one and grant the Manager permission to the user.

Reviewer: ```suggestion
If you have the `User` role assigned, you can manage your own account and change the language. To manage space group members and space member roles, an Admin or Space Admin must create one and grant the Manager permission to the user.
```

Reviewer: Not sure what the last sentence is supposed to say. What needs to be created, a space or a user? Is there such a thing as space groups?

Guest::
Users with the `Guest` role assigned can only change their language.

== Space Roles and Permissions

// https://github.com/owncloud/web/blob/3ad8561a1c7ac5ea4f1b30397569f4b9c18166f7/packages/web-pkg/src/services/permissionManager.ts

When it comes to permissions for a space, three roles are available:

* Manager
* Editor
* Viewer

Note that the creator of a space becomes manager automatically, can add other users and delegate his Manager role to one or more space members.

The following table describes the space roles and their permissions. Note that permissions are added the higher the role gets.

[width=60,cols=".^30%,100%",options="header",]
|===
| Role
| Permissions

| Viewer
a| * read documents,
* view images,

> * play videos,

Reviewer 2: download is not mentioned - is that assumed/implied by "read documents"?

> The following table describes the Space Roles and it's permissions. Note that permissions are added as higher the role gets.

Reviewer: ```suggestion
The following table describes the space roles and its permissions. Note that permissions are added the higher the role gets.
```

* get info about the space.

| Editor
a| * modify documents and even delete them,

> * upload content to a space,

Reviewer 2: do we want to mention move/rename?
the ability to download, delete and upload kind of implies that the holder has the ability to rename or move a resource, because rename or move can always be done the inefficient way anyhow, b y downloading the resource, deleting it, then upload to the new location/name.

* edit the space name,
* edit the space description,
* add or change the image for the space,
* empty the trash bin of the space.

| Manager
a| * invite additional members,
* share a space,
* deactivate and activate a space ^(1)^,
* delete a space
* change the roles of members, +
including those of other managers. A downgrade to viewer or editor is possible. 
|===

(1) ... With deactivating, a space Manager can disable any access to a space including sharing and syncing. This can be used to keep data in a frozen state before finally deleting it later on. To regain full access including sharing and syncing, a deactivated space can be activated again.

=== Special Cases

> When a space has only one user with the Manager role, it may happen that the space becomes unmanaged. Possibly because the user does not have access anymore like when leaving the company or when serious health conditions occur making any change impossible. To solve this issue and for the time being, you must incorporate that user at your _external_ IDP side and delegate the Manager role to another space member post login.

Reviewer: ```suggestion
When a space has only one user with the Manager role, it may happen that the space becomes unmanaged. Possibly because the user with that role leaves the company or suffers serious health issues. To solve this problem for the time being, you must incorporate that user at your _external_ IDP side and delegate the Manager role to another space member post login.
```
